pallets/asset/src/identifier.rs

```diff
@@ -0,0 +1,176 @@
+use codec::{Decode, Encode};
+use core::convert::TryInto;
+use sp_std::vec::Vec;
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub enum Identifier {
+    CUSIP([u8; 9]),
+    CINS([u8; 9]),
+    ISIN([u8; 12]),
+    LEI([u8; 20]),
+    EMPTY,
+}
+
+impl Default for Identifier {
+    fn default() -> Self {
+        Identifier::EMPTY
+    }
+}
+
+impl Identifier {
+    pub fn cusip(bytes: [u8; 9]) -> Option<Identifier> {
+        (cusip_checksum(&bytes[..8]) == bytes[8] - b'0').then_some(Identifier::CUSIP(bytes))
+    }
+
+    pub fn cins(bytes: [u8; 9]) -> Option<Identifier> {
+        (cusip_checksum(&bytes[..8]) == bytes[8] - b'0').then_some(Identifier::CINS(bytes))
+    }
+
+    pub fn isin(bytes: [u8; 12]) -> Option<Identifier> {
+        let (s1, s2) = bytes
+            .iter()
+            .map(|b| byte_value(*b))
```

Comment:
Use `.copied` and eta reduce here as well

Reply:
eta reduce?

Reply:
TL;DR: Eta reduction means that you take a lambda term (or in Rust, more or less a closure), e.g., `|x| function(x)`and simplify this to `function`. So in this case, you'd do (assuming we've first `.copied()`):

```suggestion
            .map(byte_value)
```
instead of:
```rust
            .map(|b| byte_value(b))
```

-------------------

Links and elaboration:

In terms of the lambda calculus, eta reduction is about taking the term `\x -> f x`, where, if we assume STLC (Simply Typed Lambda Calculus), we have `f : A -> B`, `x : A`, and `(\x -> f x) : A -> B` and transforming it to just `f : A -> B`.

- https://sookocheff.com/post/fp/eta-conversion/
- https://wiki.haskell.org/Eta_conversion